Adding basic WooAI plugin with product description generation

joelclimbsthings opened the pull request:
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Adding the basic structure of a plugin meant to capture AI experiments in WooCommerce, located under `plugins/woo-ai`.

In addition, this PR contains the necessary changes to implement a basic version of this product description experiment (pcQJnU-1sq-p2)
 There are a few changes:
- I've only included the basic functionality, leaving out the iterative "make shorter...etc" options. We may want to rethink this flow, and I wanted to get this out the door sooner rather than later.
- Converted _most_ of the relevant jQuery code to React, with an exception or two that can be addressed in a follow-up.

This also adds basic functionality for a rest API for GPT style text generation. A couple notes on the API:

**Url**: `/wp-json/wooai/text-generation`
**It accepts a few parameters:**
- `prompt` - Basic prompt to deliver to the GPT service 
- `temperature` - Sets the temperature parameter in the GPT call.
- `previous_messages` - An array of all messages that have occurred in the conversation.
- `model` - Identifier for the model to use.

**A few next steps:**
- Flesh out model abstraction, likely with a class for each model (ie openai would be one) and remove open-ai specific code from basic handler.
- Any naming updates (didn't give a lot of thought to this)
- Convert remaining jQuery code to React
- Fix mono-repo tooling (`build` isn't being run when it should)
- Gracefully fail on errors, when API keys are missing / etc.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Checkout branch and build monorepo
2.  Run the `build` operation for the WooAi plugin by running `pnpm run build` within the plugin directory.
3.  Add the open-ai constant to your local, defining it like so:
`define('OPEN_AI_KEY', 'KEY_HERE');`

This is needed for the API calls to work.

4. Link the plugin to your local, and Activate it. 
5. Navigate to Product -> Add new, and click the `Write it for me` button.

<!-- End testing instructions -->

joelclimbsthings committed:
Adding basic wooai plugin files

joelclimbsthings committed:
Nearly working product description experiment

joelclimbsthings committed:
Adding API basics

joelclimbsthings committed:
Working insecure prototype

joelclimbsthings committed:
Converting curl methods to wordpress handlers

joelclimbsthings committed:
Cleaning up some obsolete code

joelclimbsthings committed:
Adding support for previous_messages and temperature

joelclimbsthings committed:
Adding changelog

joelclimbsthings committed:
Correcting dependency version

joelclimbsthings committed:
Cleaning up references

joelclimbsthings committed:
Adding api key to models

joelclimbsthings committed:
Changing api base

tommyshellberg commented:
In case it helps others with setting everything up, I added a few things to /plugins/woocommerce/.wp-env.json.

This will automatically load and activate WC + Woo AI plugin when starting the environment with `pnpm -- wp-env start`:
```
    "plugins": [
        "../../plugins/woo-ai",
        "../woocommerce"
    ],
```

You can add the API KEY to the "config" area, like so:

```
    "config": {
        "JETPACK_AUTOLOAD_DEV": true,
        "WP_DEBUG_LOG": true,
        "WP_DEBUG_DISPLAY": true,
        "ALTERNATE_WP_CRON": true,
        "OPEN_AI_KEY": "sk-xxxxxxxxxxxxxx"
    },
````

tommyshellberg commented on plugins/woo-ai/src/product-description/product-description-form.tsx:
```diff
@@ -0,0 +1,173 @@
+					onClick={ async () => {
+						try {
+							setFetching( true );
+							const response = await apiFetch< WooApiResponse >( {
```
This seems to be the wrong path, should it be `/wooai/text-generation`?

joelclimbsthings replied:
Whoops! Last minute change, and forgot that bit. Fixed in https://github.com/woocommerce/woocommerce/pull/38309/commits/9bc36a16740f204bbc7cb5b107d8aad2c5e84149 👍🏻

tommyshellberg commented on plugins/woo-ai/api/text-generation/class-text-generation.php:
```diff
@@ -0,0 +1,172 @@
+	 */
+	protected $models = array(
+		'openai' => array(
+			'url' => 'https://api.openai.com/v1/chat/completions',
```
I think we'll want to try using the OpenAI endpoints on WPCOM where possible rather than calling OpenAI endpoints directly. I'm not sure what auth methods we need or if its possible to do with a site outside of the WPCOM environment, though?

joelclimbsthings replied:
I definitely agree, although given the uncertainty we might want to do this as a follow-up issue so we can get a starting point merged.

tommyshellberg replied:
Added a follow-up task. 👍

tommyshellberg commented on plugins/woo-ai/api/text-generation/class-text-generation.php:
```diff
@@ -0,0 +1,172 @@
+			)
+		);
+
+		if ( is_wp_error( $raw_response ) ) {
```
Do you think we should check for a response code too? 
For instance, we may get a `429` response because OpenAI is overloaded, in which case we'd want them to try again later. In this case, we'd probably want a separate condition and message to indicate to try again.

joelclimbsthings replied:
Good call. I've added a simple version of this in https://github.com/woocommerce/woocommerce/pull/38309/commits/d9ca739dda56878a07d9f164d057b7f6d368eb03 , although this might be influenced by the WPCOM APIs when we switch over to those.

joelclimbsthings committed:
Fixing api endpoint

joelclimbsthings committed:
Fixing php linting warning

joelclimbsthings committed:
Forwarding bad status codes

joelclimbsthings committed:
Minor refactor

tommyshellberg commented on plugins/woo-ai/api/text-generation/class-text-generation.php:
```diff
@@ -0,0 +1,179 @@
+	}
+
+	/**
+	 * Check if a given request has access to read.
```
*has access to create*?

tommyshellberg commented on plugins/woo-ai/api/text-generation/class-text-generation.php:
```diff
@@ -0,0 +1,179 @@
+	}
+
+	/**
+	 * Get the onboarding tasks.
```
Some leftovers from copy pasta: *Get the text completion*

tommyshellberg commented on plugins/woo-ai/api/text-generation/class-text-generation.php:
```diff
@@ -0,0 +1,179 @@
+		$prompt            = $request->get_param( 'prompt' );
+		$temperature       = $request->get_param( 'temperature' ) ? $request->get_param( 'temperature' ) : 1;
+		$previous_messages = $request->get_param( 'previous_messages' ) ? $request->get_param( 'previous_messages' ) : array();
+		$request_model     = $request->get_param( 'model' ) ? $request->get_param( 'model' ) : 'openai';
```
model doesn't actually look like an allowed param for this endpoint, should we leave this out for now and just set it to `openai` to reduce complexity?

joelclimbsthings replied:
Sounds good, simplified in https://github.com/woocommerce/woocommerce/pull/38309/commits/f53bde54458a34492beea69974d67f1704c7fc0c.

tommyshellberg commented on plugins/woo-ai/api/text-generation/class-text-generation.php:
```diff
@@ -0,0 +1,179 @@
+
+		$post_data = array(
+			'messages'    => $messages,
+			'model'       => 'gpt-3.5-turbo',
```
It seems like this model (gpt 3.5) could be tied to a key within one of the models of $this->models rather than hard-coded here. It would be easier to switch back and forth between 4.0, for instance. 
But if that's part of refactoring the models, it's fine as it is.

joelclimbsthings replied:
Yeah, the model abstraction was 10% done and I decided to punt it to a follow-up, so we'll address this in the related issue.

joelclimbsthings committed:
Addressing feedback

joelclimbsthings commented:
Thanks @tommyshellberg ! I addressed the feedback in https://github.com/woocommerce/woocommerce/pull/38309/commits/f53bde54458a34492beea69974d67f1704c7fc0c.

tommyshellberg commented:
This looks good enough to be a base branch for us to work off of 👍 
I'm curious what we're going to do with this PR, though, since we're not going to merge into `trunk`.
Should we merge this PR into a different branch? Just leave this branch as-is and close the PR (meaning the PR was just a placeholder for the review process)?

joelclimbsthings committed:
Correcting @wordpress/env version

joelclimbsthings commented:
@tommyshellberg I actually thought we were going to have this live within the mono-repo (in trunk) ?  As per [this conversation](p1684252646499249/1684229392.229919-slack-C056QSJJQ91). We could alternatively just merge everything to a feature branch as well, although we'll get out of sync with `trunk`, and we will eventually want this to live...somewhere.

nima-karimi commented:
IMO, it's better to have everything in a feature branch and only merge to `trunk` when we are ready to release the plugin. We can keep this branch and work on it as a base feature branch.

tommyshellberg commented:
> IMO, it's better to have everything in a feature branch and only merge to `trunk` when we are ready to release the plugin. We can keep this branch and work on it as a base feature branch.

This was my gut feeling as well, particularly since this plugin hasn't been thoroughly reviewed for security issues.
